core/src/main/java/ai/timefold/solver/core/impl/heuristic/selector/move/generic/list/ruin/ListRuinRecreateMove.java

```diff
@@ -95,6 +95,19 @@ protected void doMoveOnGenuineVariables(ScoreDirector<Solution_> scoreDirector)
         }
 
         for (var entry : entityToInsertedValuesMap.entrySet()) {
+            if (!entityToOriginalPositionMap.containsKey(entry.getKey())) {
+                // The entity has not been evaluated while creating the entityToOriginalPositionMap,
+                // meaning it is a new destination entity without a ListVariableBeforeChangeAction
+                // to restore the original elements.
+                // We need to ensure the before action is executed in order to restore the original elements.
+                var beforeActionElementList = new ArrayList<>(listVariableDescriptor.getValue(entry.getKey()));
+                for (var element : entry.getValue()) {
+                    var location = listVariableStateSupply.getLocationInList(element).ensureAssigned();
+                    beforeActionElementList.set(location.index(), null);
+                }
+                recordingScoreDirector.recordBeforeListVariableChanged(listVariableDescriptor, entry.getKey(),
+                        beforeActionElementList.stream().filter(Objects::nonNull).toList());
```

Comment:
This is strange, and probably inefficient.
Why put the null entries into the list, only to remove them later?

Reply:
I'm still considering the approach. The logic confirmed that my investigation into the cause of the issue was accurate. The goal is to recompute the related entity's list after the change has been completed.

It is not ready yet.